solutions/silver/cf-104002E.mdx

```diff
@@ -0,0 +1,71 @@
+---
+id: cf-104002E
+source: CF
+title: William and Robot
+author: Alex Du
+---
+
+## Explanation
+
+First note that the only requirement for a valid solution is that for any even number $k \leq n$, William must have selected at most $\frac{k}{2}$ integers.
+
+To solve this problem, we can use a greedy algorithm.
+
+For any even $k$, if we know the $\frac{k}{2}$ integers that make the highest sum, we can use this to find the $\frac{k}{2}+1$ integers for $k+2$.
+
+The base case is $k = 2$, where the highest sum is simply the maximum of the first two integers.
+
+Similarly, to gain the highest sum for our current $k$, we will always select the maximum of the $k-1$ and $k-2$ elements, since we now have another turn and this will build directly off of our previous solution.
+
+However, it is sometimes optimal to take both integers, rather than just the maximum of the two. After $k=2$, we can choose to remove any previous integer we selected and replace it with the minimum of the two integers we are currently considering. Since we can remove any integer, we will remove the minimum. Therefore, this is only worth it if the minimum of the two integers is greater than the minimum of all the integers we already selected. We need to quickly access and remove the smallest integer, so this can be done with a priority queue or a multiset.
```

Comment:
```suggestion
However, it is sometimes optimal to take both integers, rather than just the maximum of the two. After $k=2$, we can choose to remove any previous integer we selected and replace it with the minimum of the two integers we are currently considering. Since we can remove any integer, we will remove the minimum. Therefore, this is only worth it if the minimum of the two integers is greater than the minimum of all the integers we already selected. To quickly access and remove the smallest integer, we can use either a priority queue or a multiset.
```